chainstore/chain_store.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chainstore
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/hypersdk/consts"
+
+	hcontext "github.com/ava-labs/hypersdk/context"
+)
+
+const (
+	blockPrefix         byte = 0x0 // TODO: move to flat files (https://github.com/ava-labs/hypersdk/issues/553)
+	blockIDHeightPrefix byte = 0x1 // ID -> Height
+	blockHeightIDPrefix byte = 0x2 // Height -> ID (don't always need full block from disk)
+	lastAcceptedByte    byte = 0x3 // lastAcceptedByte -> lastAcceptedHeight
+)
+
+type Config struct {
+	AcceptedBlockWindow             uint64 `json:"acceptedBlockWindow"`
+	BlockCompactionAverageFrequency int    `json:"blockCompactionFrequency"`
+}
+
+func NewDefaultConfig() Config {
+	return Config{
+		AcceptedBlockWindow:             50_000, // ~3.5hr with 250ms block time (100GB at 2MB)
+		BlockCompactionAverageFrequency: 32,     // 64 MB of deletion if 2 MB blocks
+	}
+}
+
+type ChainStore[T Block] struct {
+	config  Config
+	metrics *metrics
+	log     logging.Logger
+	db      database.Database
+	parser  Parser[T]
+}
+
+type Block interface {
+	ID() ids.ID
+	Height() uint64
+	Bytes() []byte
+}
+
+type Parser[T Block] interface {
+	ParseBlock(context.Context, []byte) (T, error)
+}
+
+func New[T Block](
+	hctx *hcontext.Context,
+	namespace string,
+	parser Parser[T],
+	db database.Database,
+) (*ChainStore[T], error) {
+	registry, err := hctx.MakeRegistry(namespace)
+	if err != nil {
+		return nil, err
+	}
+	metrics, err := newMetrics(registry)
+	if err != nil {
+		return nil, err
+	}
+	config, err := hcontext.GetConfigFromContext(hctx, namespace, NewDefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChainStore[T]{
+		config:  config,
+		metrics: metrics,
+		log:     hctx.Log(),
+		db:      db,
+		parser:  parser,
+	}, nil
+}
+
+func (c *ChainStore[T]) GetLastAcceptedHeight(_ context.Context) (uint64, error) {
+	lastAcceptedHeightBytes, err := c.db.Get([]byte{lastAcceptedByte})
```

Comment:
Seems like it might make sense to just make this byte array a `var` since we're frequently re-using it